src/views/Workflow.vue

```diff
@@ -0,0 +1,171 @@
+<template>
+  <div id="workflow-panel" class="fill-height">
+    <workflow :workflow-tree="workflowTree" />
+  </div>
+</template>
+
+<script>
+import { mixin } from '@/mixins/index'
+import { mapState, mapGetters } from 'vuex'
+import Workflow from '@/components/cylc/workflow/Workflow'
+
+// query to retrieve all workflows
+const QUERIES = {
```

Comment:
I think this query and the subscription logic can now be safely removed.

Reply:
I think we still need this query? (though I spent some two minutes wondering if you were right first).

I think you mean we already have a query in the background for `GScan`?

The `GScan` query has no `familyProxies` and no `nodesEdges`. This query in the `Workflow` view is to bring the data required for the components/widgets that can be added.

Does that make sense?

Reply:
The different views should be registering the subscriptions for the things they need. If we set the query here then we will always be requesting node-edges information even when no using the graph view.

Reply:
Yup. But I couldn't find another place to add the query.

Tree and Graph components are pure components, not bound to routes and have a different lifecycle (i.e. the beforeDestroy may not be called depending on what you are doing).

Reply:
That was not a very constructive comment. Let me try to adapt the current pattern to the tree and gtaph components and then try to move the queries over there.

It might just work 👍

Reply:
Aah, I didn't realise that the event hooks fire differently.

Perhaps we could make the subscription a `computed` element which could change based on the state of the component and make the workflow component responsible to collecting and issuing subscriptions?

Reply:
What about leaving the queries here, but the initial state of the view is no query subscribed? Then when users click on "Add View", a new query will be added based on the user selected option (tree/graph).

Reply:
I think the subscription should be controlled by the component (tree, graph etc):

* Closer association of subscription and consumption of data:
  * Possibility for auto unit-testing that consumption matches subscription.
  * Possibility to plug components into a different framework (e.g. run component as a standalone application) nice for testing/integration.
* More decentralised and plugin friendly:
  * Plugins are self-contained providing an interface to the subscription.
* More powerful allowing performance enhancements:
  * The subscription of a view is a function of its state...
  * The subscription may change during the life of the component as what the user chooses to see could effect the subscription we issue.
  * This allows for opportunistic optimisation, e.g. if the user is collapsing the view by family we only need to request `familyProxies` and can drop the subscription to `taskProxies` and with it `jobs` massively reducing load.

Reply:
It is a good practice to separate control from the components. Vue itself is focused on the view layer only ([ref from their user guide](https://vuejs.org/v2/guide/#What-is-Vue-js)). Adding control and too much state to components is considered bad.

This is exactly why we have a service in our code - Angular, for example, does MVC, while Vue does MV only if you consider the state management as model. By adding the control of state into components, you run risk of code tangling, and lose the separation of concerns.

The same points you listed are used as arguments against using Axios directly in components, and instead simplifying them and externalizing the control to service/API objects. As it is easier to test, maintain, optimize, etc. Without interfering with template, layout, etc.

My goal with this separation is to have the Tree/Graph/Dot/etc components as simple to understand and maintain as possible, following what's considered a simple approach in Vue.

Another goal I have in mind, is combining "snapshots" of Cylc workflows with JupyterLab widgets. So that I could use a notebook to detail and document what I am doing, and even render the workflow with the component - without the need of Cylc UI or screen shots. Which should be simple enough if we have a component that just needs something like

```
import Vue from 'vue'
import Tree from 'cylc/tree'

// whatever code my widget needs, or any other Web app
// then instantiate a simple Vue app
<tree :data=snapshotJsonObject />
```

Now, if the `Tree` component tries to subscribe, or try to execute GraphQL queries, etc, it won't be easy to re-use it. Even within Cylc, if we ever want to provide a simple static visualization it wouldn't work.

This is a contrived example, based on what I want, and while it is not a core feature of Cylc UI, I think this can be used as an example to demonstrate benefits of having components being simplified, receiving `props`, `slots`, etc.

But I can see your point of view too. It is way easier to write components that call ApolloClient or Axios directly, but IMO it gets harder to maintain and to write tests.

But I think we are not making good progress here, so we could perhaps discuss this in the next group VC, in the Feb workshop, or wait for others from @cylc/core to chime in?

Reply:
We really need subscriptions hammered down soon as this will block other work, I thought we had it sorted before.

> Now, if the Tree component tries to subscribe, or try to execute GraphQL queries

That's not what I was describing. I don't think the Tree component should subscribe or execute GraphQL queries at all. That's the job of the workflow service.

I do, however, think that the Tree component should provide a manifest of the information it requires (similar to other component patterns). This manifest is a GraphQL subscription string. 

The Tree component wouldn't control anything, just define its requirements. The logic to merge and execute subscriptions happens higher up in the system.

Fundamentally you cannot separate the subscription and the component, it would be like trying to separate the `package.json` file and the codebase, one defines the requirements for the other. You can try to move the subscription to another file, but it's still intrinsically linked to the component and will have to be updated with each change to the compoent.

Reply:
I was describing a system which works like this:

* The Cylc "Views" (tree/dot/graph) each provide an interface for obtaining the subscription:

```js
SUB = `
    workflows {
         id
    }
`

computed: {
  subscription: () => { return SUB }
}
```

* When we "spin up" Cylc "Views" we extract the subscription from it.

```js
$workflowService.subscribe(cylcView.subscription)
```

This can be done in the Workflow Component or Workflow Service itself.

* So the Workflow Service is in charge, it *fetches* subscriptions, merges them, executes GraphQL, syncs the central data store.

Reply:
>I do, however, think that the Tree component should provide a manifest of the information it requires (similar to other component patterns). This manifest is a GraphQL subscription string.

Ah, I see, sorry I misunderstood it. I think the `props` could be this manifesto then.

>Fundamentally you cannot separate the subscription and the component, it would be like trying to separate the package.json file and the codebase, one defines the requirements for the other. You can try to move the subscription to another file, but it's still intrinsically linked to the component and will have to be updated with each change to the compoent.

I see subscriptions as related to the view/route you are visiting. Not to the components you have on that view/route.

I could be wrong, but maybe this is a practical example of where we could use this separation. If we add a view/route in Cylc UI, to replace the old `cylc graph` command, we could re-use the same component. But instead of a GraphQL query, the `<graph />` component in the view/route would need just the dependency graph.

In this case, the view/route would not have any subscription. But the component would still work because it just needs to receive the data required via props. And in this case I think we would not have any GraphQL involved.

Reply:
>I was describing a system which works like this:
>
>- The Cylc "Views" (tree/dot/graph) each provide an interface for obtaining the subscription:

The difference being that in this case we have a Workflow view, with these three components. It may have 0 or multiple instances of each of these components (tree/dot/graph).

What I mean by view, being a Vue component bound to a route, where we have lifecycle of the route (user entered, left the route, or before the user entered the route, etc). And that normally uses multiple components that may be present in other views.

And what I mean by component, is any instance of Vue that is re-used across views, and that is not bound to a route. Having a different lifecycle, and that can also be re-used by Vue - i.e. Vue may not destroy it, to re-use for performance, which again changes its lifecycle (e.g. `beforeDestroy` and `destroy` would not be called, instead we would need to hook `updated`)

Reply:
`cylc graph` is a bit of a funny one, this needs to be done at the UIS (because we need to parse the suite.rc, form the dependency graph and use other Python code) so at the UI we can re-use the component, the subscription should actually be much the same too.

Reply:
(resolved by compromise, see comments below)